main.cpp

```diff
@@ -1564,26 +1564,31 @@ void info_guts(memory_access &raw_access) {
 }
 
 string missing_device_string(bool wasRetry) {
-    char b[256];
+    const size_t bufferLen = 512;
+    char b[bufferLen];
     if (wasRetry) {
-        strcpy(b, "Despite the reboot attempt, no ");
+        strncpy(b, "Despite the reboot attempt, no ", bufferLen);
     } else {
-        strcpy(b, "No ");
+        strncpy(b, "No ", bufferLen);
     }
-    char *buf = b + strlen(b);
+    char *bufErrorMessage = b + strnlen(b, bufferLen);
     if (settings.address != -1) {
         if (settings.bus != -1) {
-            sprintf(buf, "accessible RP2040 device in BOOTSEL mode was found at bus %d, address %d.", settings.bus, settings.address);
+            snprintf(bufErrorMessage, bufferLen, "accessible RP2040 device in BOOTSEL mode was found at bus %d, address %d.", settings.bus, settings.address);
         } else {
-            sprintf(buf, "accessible RP2040 devices in BOOTSEL mode were found with address %d.", settings.address);
+            snprintf(bufErrorMessage, bufferLen, "accessible RP2040 devices in BOOTSEL mode were found with address %d.", settings.address);
         }
     } else {
         if (settings.bus != -1) {
-            sprintf(buf, "accessible RP2040 devices in BOOTSEL mode were found found on bus %d.", settings.bus);
+            snprintf(bufErrorMessage, bufferLen, "accessible RP2040 devices in BOOTSEL mode were found found on bus %d.", settings.bus);
         } else {
-            sprintf(buf, "accessible RP2040 devices in BOOTSEL mode were found.");
+            snprintf(bufErrorMessage, bufferLen, "accessible RP2040 devices in BOOTSEL mode were found.");
         }
     }
+    if (settings.force) {
+        char* bufForceError = b + strnlen(b, bufferLen);
+        snprintf(bufForceError, bufferLen, "\nTo force a device into BOOTSEL mode, make sure the RP2040 is configured to use USB-CDC (USB stdio).");
```

Comment:
If you're using `snprintf` to prevent string-overflows, do you also need to pass in a smaller value than `bufferLen` if the string `b` already contains some content?

Reply:
True, the same holds for the `bufErrorMessage` strings above.

The length argument should be something along the lines of `bufferLen - strnlen(b, bufferLen)`. I'm not sure if this causes an off-by-one error regarding the null terminator (e.g. should it be `bufferLen - strnlen(b, bufferLen) - 1` or not)?

Reply:
> I'm not sure if this causes an off-by-one error

I dunno either :man_shrugging: I'm more of a Python programmer than a C programmer :wink: